draft-ietf-taps-interface.md

```diff
@@ -1011,6 +1011,30 @@ ChallengeCallback := NewCallback({
 SecurityParameters.SetIdentityChallengeCallback(challengeCallback)
 ~~~
 
+## Forcing the use of Specific Protocols
+
+Some application protocols need to be exercised on top of specific protocols, e.g. because they use lower layer protocol fields in their specification.
+In such cases, the application can request the transport system to use a specific set of protocols as part on the local endpoint.
+
+~~~
+LocalSpecifier := NewLocalEndpoint()
+LocalSpecifier.WithProtocol("http")
+~~~
+
+Multiple sets can be specified as alternatives:
+
+~~~
+LocalSpecifier := NewLocalEndpoint()
+LocalSpecifier.WithProtocols("tcp", "ipv4")
+LocalSpecifier.WithProtocols("tcp", "ipv6")
+~~~
+
```

Comment:
Why is this done on the local endpoint object rather than on the transport properties object?

Reply:
The reason is that really forcing a protocol end-to-end is hard and requires crypto. Therefore, we only force the protocol at the local endpoint. On the path, there may be proxies or protocol converters, so that the Remote Endpoint may use another protocol. Adding it on the Local Endpoint object makes this explicit.

Reply:
I also find using the endpoint object confusing and agree with Max that it should be a transport property. It is used as a property for the selection and racing so we should keep all this together. (I am not convinced by the argument for why we should use the endpoint object. The path could violate other transport properties as well then.)

Reply:
There is another argument why realising this as Transport Property is a little painful: what would the following mean:
~~~
transportPorperties.require(protocol, UDP)
transportPorperties.require(protocol, TCP)
transportPorperties.require(protocol, IPv4)
transportPorperties.require(protocol, IPv6)
~~~
Does it mean what I intended (require to use TCP or UDP and prefer IPv4), or is this only satisfied by TCPoverUDPoverIPv6overIPv4 or TCPoverUDPoverIPv4overIPv6?

~~~
NewLocalEndpoint.WithProtocols("udp","tcp")
NewLocalEndpoint.WithProtocols("IPv4","IPv6"),
~~~
Is clear - I want either TCP or UDP over IPv4 or IPv6 - and does not allow to use other preferences and thus reduces complexity.

Reply:
I think the example that you give here makes the endpoint case clear because you assume an implicit "or" between parameters of one call. In the other case, you arrive at something ambiguous because you have only one flat property called "protocol".

The distinction that creates the problem here, to me, isn't about the local endpoint or use of a transport property; it's about semantics of the defined call that you use (or assume).